source/Engine.cpp

```diff
@@ -1461,10 +1461,26 @@ void Engine::CalculateStep()
 
 	// Keep track of the flagship to see if it jumps or enters a wormhole this turn.
 	const Ship *flagship = player.Flagship();
+	bool flagshipWasUntargetable = (flagship && !flagship->IsTargetable());
 	bool wasHyperspacing = (flagship && flagship->IsEnteringHyperspace());
-	// Move all the ships.
+	const System *flagshipSystem = (flagship ? flagship->GetSystem() : nullptr);
+	// First, move the player's flagship.
+	MoveShip(player.FlagshipPtr());
+	bool flagshipIsUntargetable = (flagship && !flagship->IsTargetable());
+	bool flagshipBecameTargetable = flagshipWasUntargetable && !flagshipIsUntargetable;
+	// Then, move the other ships.
 	for(const shared_ptr<Ship> &it : ships)
+	{
+		if(it == player.FlagshipPtr())
+			continue;
+		bool wasUntargetable = (!it->IsTargetable());
 		MoveShip(it);
+		// If either the flagship or this ship uncloaked this frame and both are uncloaked,
+		// then they have encountered each other.
+		if(((wasUntargetable && it->IsTargetable()) || flagshipBecameTargetable)
+			&& flagshipSystem == it->GetSystem() && !flagshipIsUntargetable)
```

Comment:
The comment is somewhat pointless now since we've converted all the comparisons into variables with descriptive names.
```suggestion
		bool isTargetable = it->IsTargetable();
		if(flagshipSystem == it->GetSystem()
			&& ((wasUntargetable && isTargetable) || flagshipBecameTargetable)
			&& isTargetable && flagshipIsTargetable)
```